flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java

```diff
@@ -243,6 +246,22 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object msg) throws Exc
             if (toRelease != null) {
                 releaseViewReader(toRelease);
             }
+        } else if (msg.getClass() == PartitionRequestNotifierTimeout.class) {
+            PartitionRequestNotifierTimeout partitionRequestNotifierTimeout = (PartitionRequestNotifierTimeout) msg;
+
+            // Send partition not found message to the downstream task when the notifier is timeout.
+            final PartitionRequestNotifier partitionRequestNotifier = partitionRequestNotifierTimeout
+                    .getPartitionRequestNotifier();
+            final ResultPartitionID resultPartitionId = partitionRequestNotifier.getResultPartitionId();
+            final InputChannelID inputChannelId = partitionRequestNotifier.getReceiverId();
+            availableReaders.remove(partitionRequestNotifier.getViewReader());
+            allReaders.remove(inputChannelId);
```

Comment:
Not familiar with this mechanism. Could you explain why we need to do the removal?

Reply:
When netty server receives partition request, it will create a reader and add it to allReaders even when the partition is not registered. This reader should be removed when the partition request listener for the partition may be not registered in the TM anymore.